src/Agent.Plugins/PipelineCache/PipelineCacheServer.cs

```diff
@@ -42,32 +52,175 @@ public class PipelineCacheServer
+                var outputFileStream = File.Create(outputFile);
+                var archieveFile = Path.Combine(path, Guid.NewGuid().ToString() + "archieve.tar");
+                using(StreamWriter stream = new StreamWriter(outputFileStream))
                 {
-                    Fingerprint = fingerprint,
-                    RootId = result.RootId,
-                    ManifestId = result.ManifestId,
-                    ProofNodes = result.ProofNodes.ToArray(),
-                };
+                    foreach(var file in relativeFilesPath)
+                    {
+                        stream.WriteLine(file);
+                    }
+                    foreach(var directory in relativeDirectoryPath)
+                    {
+                        stream.WriteLine(directory);
+                    }
+                }*/
 
-                // Cache the artifact
-                PipelineCacheActionRecord cacheRecord = clientTelemetry.CreateRecord<PipelineCacheActionRecord>((level, uri, type) =>
-                    new PipelineCacheActionRecord(level, uri, type, PipelineArtifactConstants.SaveCache, context));
-                CreateStatus status = await pipelineCacheClient.CreatePipelineCacheArtifactAsync(options, cancellationToken, cacheRecord);
+                if (!isTar)
+                {
+                    //Upload the pipeline artifact.
+                    PipelineCacheActionRecord uploadRecord = clientTelemetry.CreateRecord<PipelineCacheActionRecord>((level, uri, type) =>
+                        new PipelineCacheActionRecord(level, uri, type, nameof(dedupManifestClient.PublishAsync), context));
+                    PublishResult result = await clientTelemetry.MeasureActionAsync(
+                        record: uploadRecord,
+                        actionAsync: async () =>
+                        {
+                            return await dedupManifestClient.PublishAsync(path, cancellationToken);
+                        });
 
-                // Send results to CustomerIntelligence
-                context.PublishTelemetry(area: PipelineArtifactConstants.AzurePipelinesAgent, feature: PipelineArtifactConstants.PipelineCache, record: uploadRecord);
-                context.PublishTelemetry(area: PipelineArtifactConstants.AzurePipelinesAgent, feature: PipelineArtifactConstants.PipelineCache, record: cacheRecord);
+                    CreatePipelineCacheArtifactOptions options = new CreatePipelineCacheArtifactOptions
+                    {
+                        Fingerprint = fingerprint,
+                        RootId = result.RootId,
+                        ManifestId = result.ManifestId,
+                        ProofNodes = result.ProofNodes.ToArray(),
+                    };
+
+                    // Cache the artifact
+                    PipelineCacheActionRecord cacheRecord = clientTelemetry.CreateRecord<PipelineCacheActionRecord>((level, uri, type) =>
```

Comment:
All the common functionality should not be duplicated. This is just for testing the initial end to end scenario.